### ISSUE TITLE
two models ensable cost two times than one model

### ISSUE THREAD

**Issue by phveneno** [tensorflow/tensor2tensor#560]
hi

I am try the model ensable.  And the result is +1.0 BLEU in our data set. But two models ensable cost two times when decoding.

I am using version 1.3.2.  for the easy way to implement. I use _beam_decode_slow not _fast_decode. Because _fast_decode use some caches. ( The _fast_decode also can used to do model ensable)

here is my way to implement. (a tricky way)

In function _beam_decode_slow in t2t_model.py, there is a symbols_to_logits_fn function.  it used to calculate the decoder output by given the previous decoding result.  the function self.model_fn is called once. I call self.model_fn twice, but using different variable_scope name.  the code is like this, different name stands different model.

`logits, _ = self._model_fn( features, name="model_1")`
`...`
`logits_2, _ = self._model_fn( features, name="model_2")`

the model_fn was modified like this.

`if name != None and name != "":`
`    ` `body_name = name + "/body"`

`else:`
`    ` `body_name = "body"`
 `    ...`
`with tf.variable_scope(body_name, reuse=self._problem_idx > 0):`

and in fuction beam_search in beam_search.py. I can get the decoder result of two model. Than doing the softmax and then get the mean of softmax result, the log  probability. 

but the decoding time was raise two much. I using batch_size = 1, beam=4,  the gpu is M40. 

Is there someone give some advise to accelerate this. I found no useful document for doing model ensable in tf.

thanks so much.

**Comment by rsepassi** [comment]
I'm not sure I'm understanding your question. You're ensembling 2 models and you observe that the cost of evaluation goes up 2x. Is that right? If so, that's not surprising.

**Comment by lukaszkaiser** [comment]
Indeed: that's the problem with ensembles, they are slow at inference time. One way to avoid this is to train models as if they were doing distillation. You train 2 models (might just make 2 copies in different tf.variable_scopes) and then for the loss, you nudge them to make similar predictions. But importantly, you start with different weights and wait a bit (can be done with tf.cond checking for global_step) before nudging them to do the same, so they are different. This approach, called co-distillation, seems to recover most of the benefits of ensembling (though not all) while giving you a fast model at inference time. Check out the recent paper: https://openreview.net/forum?id=rkr1UDeC-
